clients/js/src/index.ts

```diff
@@ -39,3 +39,61 @@ export class CacReader {
         return deepMerge(targetConfig, ...requiredOverrides);
     }
 }
+
+export class ExperimentReader {
+    experiments: ExperimentStore;
+
+    constructor(experiments: Experiments) {
+        this.experiments = new Map(experiments.map(exp => [exp.id, exp]));
+    }
+
+    public getApplicableVariant(data: IObject, toss: number): Array<String> {
+        const experiments = this.getSatisfiedExperiments(data);
+        const variants = [];
+        for (const exp of experiments) {
+            const v = this.decideVariant(exp.traffic_percentage, exp.variants, toss);
+            if (v) {
+                variants.push(v.id)
+            }
+        }
+        return variants;
+    };
+
+    getSatisfiedExperiments(data: IObject): Experiments {
+        const experiments = [];
+        const iter = this.experiments.values();
+
+        while (true) {
+            let result = iter.next();
+            if (result.done) break;
+            if (jsonLogic.apply(result.value.context, data)) {
+                experiments.push(result.value);
+            }
+        }
+
+        return experiments;
+    };
+
+    // decide which variant to return among all applicable experiments
+    decideVariant(
+        traffic: number,
+        applicable_variants: Variants,
+        toss: number,
+    ): Variant | undefined {
+        if (toss < 0) {
+            for (const variant of applicable_variants) {
+                if (variant.variant_type == VariantType.EXPERIMENTAL) {
+                    return variant;
+                }
+            }
+        }
+        const variant_count = applicable_variants.length;
+        const range = traffic * variant_count;
+        if (toss >= range) {
+            return undefined;
+        }
+        let buckets = Array.apply(null, Array(variant_count)).map(function (_, i) { return traffic * (i + 1); });
```

Comment:
This is fine, but too computations happening here, making an array of `NULL`s, mapping over it, and then in the finding the index of the bucket and finally using the index to get the variant/bucket.
I think this can be just written in one single `for` loop.